src/Drips.sol

```diff
@@ -307,78 +310,28 @@ library Drips {
-    struct DefaultEnd {
-        uint32 start;
-        uint136 amtPerSec;
-    }
-
-    /// @notice Adds a `DefaultEnd` to the list of receivers while keeping it sorted.
-    /// @param defaults The list of default ends, must be sorted by start
-    /// @param length The length of the list
-    /// @param start The start time of the added receiver
-    /// @param amtPerSec The amtPerSec of the added receiver
-    /// @return newLength The new length of the list
-    function _addDefaultEnd(
-        DefaultEnd[] memory defaults,
-        uint8 length,
-        uint32 start,
-        uint128 amtPerSec
-    ) private pure returns (uint8 newLength) {
-        for (uint8 i = 0; i < length; i++) {
-            DefaultEnd memory defaultEnd = defaults[i];
-            if (defaultEnd.start == start) {
-                defaultEnd.amtPerSec += amtPerSec;
-                return length;
-            }
-            if (defaultEnd.start > start) {
-                // Shift existing entries to make space for inserting the new entry
-                for (uint8 j = length; j > i; j--) {
-                    defaults[j] = defaults[j - 1];
-                }
-                defaults[i] = DefaultEnd(start, amtPerSec);
-                return length + 1;
-            }
-        }
-        defaults[length] = DefaultEnd(start, amtPerSec);
-        return length + 1;
-    }
-
-    /// @notice Calculates the end time of drips without duration.
-    /// @param defaults The list of default ends, must be sorted by start
-    /// @param length The length of the list
-    /// @param balance The balance available for drips without duration
-    /// @return end_ The end time of drips without duration
-    function _receiversDefaultEnd(
-        DefaultEnd[] memory defaults,
-        uint8 length,
-        uint128 balance
-    ) private pure returns (uint32 end_) {
-        uint32 lastStart = 0;
-        uint136 amtPerSec = 0;
-        uint136 end = type(uint136).max;
-        for (uint8 i = 0; i < length; i++) {
-            DefaultEnd memory defaultEnd = defaults[i];
-            uint32 start = defaultEnd.start;
-            if (start >= end) break;
-            balance -= uint128(amtPerSec * (start - lastStart));
-            lastStart = start;
-            amtPerSec += defaultEnd.amtPerSec;
-            end = start + (balance / amtPerSec);
+        // no default drips
+        if (defaultDivisor == 0) {
+            return 0;
         }
-        if (end > type(uint32).max) end = type(uint32).max;
-        return uint32(end);
+        // main formular to calculate default end time
+        // a ...  amtPerSec
+        // s ...  start
+        // defaultEnd =(balance + a_1*s_1 + ... + a_n*s_n)/(a_1 + ... a_n)
+        uint136 defaultEnd = uint136((balance + defaultSum) / defaultDivisor);
+        if (defaultEnd > type(uint32).max) defaultEnd = type(uint32).max;
+        require(defaultEnd >= highest, "Run out of funds before default drips start");
```

Comment:
I had it first like that.  Some of the existing unit tests failed. 
In general I agree with you. I will investigate it.

Reply:
I investigate why so many tests failed, when I changed that.

It is related to how we are performing the `assertBalance` check in the Drips test.

Image a case  we have a defaultDrip with 3 amtPerSec and we call the `assertBalance` check after the user hasn't enough balance to cover it anymore. 

Let's say the leftover balance is `1`.

The `assertBalance` basically results in a `setDrips` call where we pass the currReceiver as newReceiver. BalanceDelta=0. However, if it contains a defaultDrips with start=0 and duration=0. It will be interpreted as a new one. 

Then we end up the scenario where we don't have enough funds to cover the amtPerSec and the defaultEnd will in current `block.timestamp`. 

We could tolerate this user behavior. I think it would also happen if the user just wants to withdraw his balance but just passes his previous only defaultDrips. 

This would fix it.

```solidity
        require(defaultEnd > defaultsHighestStart || defaultEnd == block.timestamp, "Run out of funds before default drips start");
     
```
The alternative would be that we revert in such cases and adjust the tests.

wdyt @CodeSandwich

Reply:
I don't think that we can tolerate that scenario. How to describe this API behavior? "You must provide enough balance to cover at least 1 second of each unbound drips OR you must provide so little balance that you can't cover a single second of total amtPerSec of all unbound drips".

Another problem is that `defaultEnd != block.timestamp` if the uncovered unbound drip has a `start` in the future.

I think that we should adjust the tests. Maybe it'd be a good idea to add a balance calculating function to Drips lib and expose it in DripsHub? I think that it could be very useful for on-chain API users. Otherwise they'd need to duplicate drips logic on their side.
```solidity
function balanceAt(Storage storage s, uint256 userId, DripsReceiver[] memory receivers, uint32 timestamp) internal view returns (uint128 balance) { ... }
```

Reply:
Change to `defaultEnd > defaultsHighestStart `.
Thanks for adding #163